### PR TITLE
Count and CountA implementations

### PR DESCRIPTION
**Pull request by anderson-joyle**
Issue #284

### CODE REVIEW

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/CountFunctions.txt [hunk 1]
```diff
@@ -27,3 +27,27 @@ Errors: Error 22-25: Expected boolean. We expect a boolean (true/false) at this
...
+3
+
+>> CountA(Sort([-2, -1, 0, 1, 2], 1 / Value))
+#Error
+
+>> CountA(LastN([1,2,3,4], Blank()))
+Blank()
+
+>> CountA([1, 2, 1/0, 4, 5])
+5
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/CountFunctions.txt:47]: Can you double check this one in Power apps (with error handling feature enabled - see comment at top of #284). 

If any element is an error, it should return error.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/CountFunctions.txt [hunk 2]
```diff
@@ -27,3 +27,27 @@ Errors: Error 22-25: Expected boolean. We expect a boolean (true/false) at this
 
 >> CountIf([1, 2, 3, 4], "abc")
 Errors: Error 22-27: Expected boolean. We expect a boolean (true/false) at this point in the formula.
+
+>> CountA([1,2,3,Blank()])
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/CountFunctions.txt:31]: >>> [](http://example.com/codeflow?start=0&length=2)

nit - can you put these in their own .txt file? (We should have split up CountRows, CountIf into their own files too)

Comment (anderson-joyle) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/CountFunctions.txt]: Done.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/CountFunctions.txt (hunk 1 above)
Comment (anderson-joyle) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/CountFunctions.txt]: Thanks for clarifying. In PApps both Count and CountA will display an error message. This has been taken into account.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs [hunk 3]
```diff
@@ -136,6 +136,66 @@ public static FormulaValue CountRows(IRContext irContext, TableValue[] args)
...
+        public static FormulaValue Count(IRContext irContext, TableValue[] args)
+        {
+            var arg0 = args[0];
+            var count = 0;
+
+            var errors = new List<ErrorValue>();
+
+            foreach (var row in arg0.Rows)
+            {
+                if (row.Value.GetField(TableValue.ValueName) is ErrorValue error)
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs:149]: >row.Value [](http://example.com/codeflow?start=20&length=9)

`row.Value` may be null if the row is blank or an error. Examples:

- Count(Table({a:1},{a:2},If(Sqrt(-1)<1,{a:3}),{a:4}) // error row
- Count(Table({a:1},{a:2},If(1<0,{a:3}),{a:4}) // blank row

**Comment by CarlosFigueiraMSFT** [comment]
> >> CountRows([1, 2, 3, 4, 5])

Can you please rename this file to CountRows.txt (with 's' at the end)?

---
Refers to: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/CountRow.txt:1 in 74445e4. [](commit_id = 74445e43fe8166ec662fc79dbd0bfe7b3246b92d, deletion_comment = False)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs (hunk 3 above)
Comment (anderson-joyle) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs]: Thanks for making me aware of these scenarios.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs [hunk 4]
```diff
@@ -136,6 +136,90 @@ public static FormulaValue CountRows(IRContext irContext, TableValue[] args)
...
+        public static FormulaValue Count(IRContext irContext, TableValue[] args)
+        {
+            var arg0 = args[0];
+            var count = 0;
+
+            var errors = new List<ErrorValue>();
+
+            foreach (var row in arg0.Rows)
+            {
+                if (row.Value is null)
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs:149]: `if (row.IsBlank) { continue; }... if (row.IsError) { errors.Add(row.Error); continue; }

Comment (anderson-joyle) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs:149]: The test case **Count([1, 2, 1/0, 4, 5])**, row **1/0** is giving me the following results:

- Blank: null
- Error: null
- IsBlank: false
- IsError: false
- IsValue: true
- Value: ErrorType

Is this a valid value?

Comment (anderson-joyle) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs:149]: Please let me know so I can apply the suggested changes.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Count.txt [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿>> Count([0,""])
+1
+
+>> Count([1, 2, 1/0, 4, 5])
+#Error
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Count.txt:5]: `#Error(Kind=Div0)`

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs (hunk 4 above)
Comment (CarlosFigueiraMSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs]: Can you clarify where (in which file) you see that object? Typically we shouldn't have a DValue<T> with IsValue = true, but where the Value is an error value

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs [hunk 6]
```diff
@@ -245,6 +245,26 @@ internal static partial class Library
...
+                StandardErrorHandling<TableValue>(
+                    expandArguments: NoArgExpansion,
+                    replaceBlankValues: DoNotReplaceBlank,
+                    checkRuntimeTypes: ExactValueTypeOrBlank<TableValue>,
+                    checkRuntimeValues: DeferRuntimeValueChecking,
+                    returnBehavior: ReturnBehavior.ReturnBlankIfAnyArgIsBlank,
+                    targetFunction: Count)
+            },
             {
                 BuiltinFunctionsCore.CountRows_UO,
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:269]: [minor] Can you please move this up before CountIf, to keep the functions sorted?

**Comment by anderson-joyle** [comment]
@CarlosFigueiraMSFT I was double-checking how Excel handles error value cells in the CountA formula. It does include the error values in the counting. I'm committing the changes since CountIf was implemented also returning an error in case of an error value row. Let me know your thoughts to I can act accordingly.

**Comment by MikeStall** [comment]
> >> Count([0,""])

I just ran all these against JS to confirm and they pass. 


---
Refers to: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Count.txt:1 in 2aece30. [](commit_id = 2aece300ba47c9dada0760b0575bd45a4927030f, deletion_comment = False)

**Review by MikeStall — APPROVED** [review]

**Comment by anderson-joyle** [comment]
@MikeStall is there are manual on how I can perform JS tests from my end?

**Comment by MikeStall** [comment]
> @MikeStall is there are manual on how I can perform JS tests from my end?

Not yet ... they're some internal tests. We're working on getting them public too.

**Pull request merged**